src/server/scripts/Kalimdor/CavernsOfTime/BattleForMountHyjal/boss_archimonde.cpp

```diff
@@ -88,707 +88,473 @@ enum Events
+                    Unit::DealDamage(me, me, me->GetHealth(), nullptr, DIRECT_DAMAGE, SPELL_SCHOOL_MASK_NORMAL, nullptr, false);
                 }
-        }
+                Position wispPosition = { me->GetPositionX() + float(rand() % WISP_OFFSET), me->GetPositionY() + float(rand() % WISP_OFFSET), me->GetPositionZ(), 0.0f };
+                if (Creature* wisp = me->SummonCreature(CREATURE_ANCIENT_WISP, wispPosition))
+                {
+                    wisp->AI()->DoCast(me, SPELL_ANCIENT_SPARK);
+                    ++_wispCount;
+                }
+            }, 1500ms);
+            ScheduleTimedEvent(1500ms, [&]
+            {
+                DoCastVictim(SPELL_RED_SKY_EFFECT);
+                DoCastVictim(SPELL_HAND_OF_DEATH);
+            }, 3s);
+            ScheduleTimedEvent(2500ms, [&]
+            {
+                if (!(me->GetVictim() && me->IsWithinMeleeRange(me->GetVictim())))
+                {
+                    DoCastRandomTarget(SPELL_FINGER_OF_DEATH);
+                }
+            }, 3500ms);
+        });
+    }
 
-        void JustEngagedWith(Unit* /*who*/) override
+    void DoAction(int32 action) override
+    {
+        switch (action)
         {
-            me->InterruptSpell(CURRENT_CHANNELED_SPELL);
-            Talk(SAY_AGGRO);
-            DoZoneInCombat();
-
-            instance->SetData(DATA_ARCHIMONDE, IN_PROGRESS);
-            events.ScheduleEvent(EVENT_SPELL_AIR_BURST, urand(25000, 35000));
-            events.ScheduleEvent(EVENT_SPELL_DOOMFIRE, urand(10000, 20000));
-            events.ScheduleEvent(EVENT_SPELL_FEAR, 42000);
-            events.ScheduleEvent(EVENT_SPELL_GRIP_OF_THE_LEGION, 2000);
-            events.ScheduleEvent(EVENT_SPELL_FINGER_OF_DEATH, 1000);
-
-            instance->SetData(DATA_SPAWN_WAVES, 1);
+            case ACTION_BECOME_ACTIVE_AND_CHANNEL:
+                me->SetReactState(REACT_AGGRESSIVE);
+                me->SetVisible(true);
+                if (!_isChanneling)
+                {
+                    if (Creature* nordrassil = me->SummonCreature(CREATURE_CHANNEL_TARGET, nordrassilPosition, TEMPSUMMON_TIMED_DESPAWN, 1200000))
+                    {
+                        nordrassil->SetUnitFlag(UNIT_FLAG_NOT_SELECTABLE);
+                        nordrassil->SetDisplayId(11686); //TODO: make enum
+                        DoCast(nordrassil, SPELL_DRAIN_WORLD_TREE);
+                        _isChanneling = true;
+                        nordrassil->AI()->DoCast(me, SPELL_DRAIN_WORLD_TREE_2, true);
+                    }
+                }
+                break;
         }
+    }
 
-        void KilledUnit(Unit* victim) override
+    void DoCastProtection()
+    {
+        if (Map* map = me->GetMap())
         {
-            Talk(SAY_SLAY);
-
-            if (victim && victim->GetTypeId() == TYPEID_PLAYER)
-                GainSoulCharge(victim->ToPlayer());
+            map->DoForAllPlayers([&](Player* player)
+            {
```

Comment:
```
+ if (me->GetVictim() != player)
+     continue;
```

Reply:
![image](https://github.com/azerothcore/azerothcore-wotlk/assets/18535853/7de10b38-bd26-43a2-93c3-405fb1b2d5f0)

Reply:
Wdym doforallplayers only iterates over players by definition

Reply:
Have you tested it?

Reply:
I don't understand what you're getting at.

![image](https://github.com/azerothcore/azerothcore-wotlk/assets/83884799/eacc2bf6-7187-49fe-82ff-e06351102b21)
GetPlayers() gets all players, and not other creatures

Reply:
OK..